File: api/lib/utils/text.mjs
export function normalizeLanguage(value) {
  return String(value).trim().toLowerCase();
}

export function getQueryValue(value) {
  if (Array.isArray(value)) {
    return value[0] || "";
  }

  return value || "";
}

export function pickFirstNonEmptyQueryValue(...values) {
  for (const value of values) {
    const resolved = String(getQueryValue(value)).trim();

    if (resolved) {
      return resolved;
    }
  }

  return getQueryValue(values.at(-1));
}

export function escapeXml(value) {
  return String(value)
    .replaceAll("&", "&amp;")
    .replaceAll("<", "&lt;")
    .replaceAll(">", "&gt;")
    .replaceAll('"', "&quot;")
    .replaceAll("'", "&apos;");
}

export function truncateText(value, maxLength) {
  const text = String(value);

  if (text.length <= maxLength) {
    return text;
  }

  if (maxLength <= 3) {
    return text.slice(0, maxLength);
  }

  return `${text.slice(0, maxLength - 3)}...`;
}

export function wrapText(value, maxChars, maxLines = 2) {
  const text = String(value).trim();

  if (!text) {
    return [];
  }

  const words = text.split(/\s+/);
  const lines = [];
  let currentLine = "";

  for (let index = 0; index < words.length; index += 1) {
    const word = words[index];
    const nextLine = currentLine ? `${currentLine} ${word}` : word;

    if (nextLine.length <= maxChars) {
      currentLine = nextLine;
      continue;
    }

    if (!currentLine) {
      currentLine = truncateText(word, maxChars);
    }

    lines.push(currentLine);

    if (lines.length === maxLines - 1) {
      const remaining = words.slice(index).join(" ");
      lines.push(truncateText(remaining, maxChars));
      return lines;
    }

    currentLine = word.length > maxChars ? truncateText(word, maxChars) : word;
  }

  if (currentLine) {
    lines.push(currentLine);
  }

  return lines.slice(0, maxLines);
}

export function estimateCharCapacity(availableWidth, fontSize, factor = 0.57) {
  return Math.max(4, Math.floor(availableWidth / (fontSize * factor)));
}

function getGlyphWidthFactor(character) {
  if (character === " ") {
    return 0.36;
  }

  if (/[.,:;!|]/.test(character)) {
    return 0.28;
  }

  if (/[-_/\\()[\]{}]/.test(character)) {
    return 0.42;
  }

  if (character === "@") {
    return 0.92;
  }

  if (/[MW]/.test(character)) {
    return 1.02;
  }

  if (/[A-Z]/.test(character)) {
    return 0.84;
  }

  if (/[mw]/.test(character)) {
    return 0.9;
  }

  if (/[iljtfr]/.test(character)) {
    return 0.5;
  }

  if (/\d/.test(character)) {
    return 0.72;
  }

  return 0.72;
}

export function estimateTextWidth(value, fontSize, factor = 0.88) {
  return Array.from(String(value)).reduce(
    (total, character) =>
      total + fontSize * factor * getGlyphWidthFactor(character),
    0,
  );
}

export function truncateTextToWidth(value, maxWidth, fontSize, factor = 0.88) {
  const text = String(value);

  if (estimateTextWidth(text, fontSize, factor) <= maxWidth) {
    return text;
  }

  const ellipsis = "...";
  const ellipsisWidth = estimateTextWidth(ellipsis, fontSize, factor);

  if (ellipsisWidth >= maxWidth) {
    return ellipsis;
  }

  let output = "";

  for (const character of text) {
    const nextValue = `${output}${character}`;

    if (
      estimateTextWidth(nextValue, fontSize, factor) + ellipsisWidth >
      maxWidth
    ) {
      break;
    }

    output = nextValue;
  }

  return output ? `${output}${ellipsis}` : ellipsis;
}

export function formatPercentage(value) {
  return `${value
    .toFixed(2)
    .replace(/\.00$/, "")
    .replace(/(\.\d)0$/, "$1")}%`;
}

File: internal/util/text.go
package util

import (
	"fmt"
	"regexp"
	"strings"
)

var whitespacePattern = regexp.MustCompile(`\s+`)

func NormalizeLanguage(value string) string {
	return strings.ToLower(strings.TrimSpace(value))
}

func EscapeXML(value string) string {
	replacer := strings.NewReplacer(
		"&", "&amp;",
		"<", "&lt;",
		">", "&gt;",
		`"`, "&quot;",
		"'", "&apos;",
	)

	return replacer.Replace(value)
}

func TruncateText(value string, maxLength int) string {
	runes := []rune(value)
	if len(runes) <= maxLength {
		return value
	}

	if maxLength <= 3 {
		return string(runes[:maxLength])
	}

	return string(runes[:maxLength-3]) + "..."
}

func WrapText(value string, maxChars, maxLines int) []string {
	text := strings.TrimSpace(value)
	if text == "" {
		return nil
	}

	words := whitespacePattern.Split(text, -1)
	lines := make([]string, 0, maxLines)
	currentLine := ""

	for index := 0; index < len(words); index++ {
		word := words[index]
		nextLine := word
		if currentLine != "" {
			nextLine = currentLine + " " + word
		}

		if len([]rune(nextLine)) <= maxChars {
			currentLine = nextLine
			continue
		}

		if currentLine == "" {
			currentLine = TruncateText(word, maxChars)
		}

		lines = append(lines, currentLine)
		if len(lines) == maxLines-1 {
			remaining := strings.Join(words[index:], " ")
			lines = append(lines, TruncateText(remaining, maxChars))
			return lines
		}

		if len([]rune(word)) > maxChars {
			currentLine = TruncateText(word, maxChars)
			continue
		}

		currentLine = word
	}

	if currentLine != "" {
		lines = append(lines, currentLine)
	}

	if len(lines) > maxLines {
		return lines[:maxLines]
	}

	return lines
}

func EstimateCharCapacity(availableWidth, fontSize, factor float64) int {
	return max(4, int(mathFloor(availableWidth/(fontSize*factor))))
}

func EstimateTextWidth(value string, fontSize, factor float64) float64 {
	total := 0.0
	for _, character := range value {
		total += fontSize * factor * glyphWidthFactor(character)
	}

	return total
}

func TruncateTextToWidth(value string, maxWidth, fontSize, factor float64) string {
	if EstimateTextWidth(value, fontSize, factor) <= maxWidth {
		return value
	}

	ellipsis := "..."
	ellipsisWidth := EstimateTextWidth(ellipsis, fontSize, factor)
	if ellipsisWidth >= maxWidth {
		return ellipsis
	}

	output := strings.Builder{}
	for _, character := range value {
		nextValue := output.String() + string(character)
		if EstimateTextWidth(nextValue, fontSize, factor)+ellipsisWidth > maxWidth {
			break
		}

		output.WriteRune(character)
	}

	if output.Len() == 0 {
		return ellipsis
	}

	return output.String() + ellipsis
}

func FormatPercentage(value float64) string {
	formatted := fmt.Sprintf("%.2f", value)
	formatted = strings.TrimSuffix(formatted, ".00")
	if strings.HasSuffix(formatted, "0") && strings.Contains(formatted, ".") {
		formatted = strings.TrimSuffix(formatted, "0")
	}

	return formatted + "%"
}

func glyphWidthFactor(character rune) float64 {
	switch {
	case character == ' ':
		return 0.36
	case strings.ContainsRune(".,:;!|", character):
		return 0.28
	case strings.ContainsRune("-_/\\()[]{}", character):
		return 0.42
	case character == '@':
		return 0.92
	case strings.ContainsRune("MW", character):
		return 1.02
	case character >= 'A' && character <= 'Z':
		return 0.84
	case strings.ContainsRune("mw", character):
		return 0.9
	case strings.ContainsRune("iljtfr", character):
		return 0.5
	case character >= '0' && character <= '9':
		return 0.72
	default:
		return 0.72
	}
}

func mathFloor(value float64) float64 {
	if value < 0 {
		return float64(int(value - 1))
	}

	return float64(int(value))
}
